emma-language/src/main/scala/org/emmalanguage/compiler/lang/TreeNorm.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright © 2014 TU Berlin (emma@dima.tu-berlin.de)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.emmalanguage
+package compiler.lang
+
+import compiler.Common
+import core.Core
+
+import scala.collection.mutable
+
+
+trait TreeNorm extends Common {
+  self: Core =>
+
+  import UniverseImplicits._
+
+  /**
+   *  Starting at the return value of a suntax tree, resorts val definitions within the tree in a depth-first manner.
```

Comment:
Instead of "syntax tree", please write "let-block".

Also, please elaborate on the comment. Something like the following:
-----------------
The order of ValDefs will be a post-order traversal of the DAG that has ValDefs as its nodes, and there is a directed edge from u to v if the RHS of u's ValDef refers to v's ValDef.

Note that the order of the recursive calls from a single node is determined by the order in which the `collect` call on the RHS returns the ValRefs. For this reason, this has to be a bottom-up transform, so that if there is a let-block in the RHS of a ValDef then it should have already been normalized.
-----------------